state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,447 @@
+                        .subscription_id_to_event_subscription
+                        .entry(*subscription_id)
+                    {
+                        event_subscription.get_mut().buffer_event(event.clone());
+                        event_subscription_ids_to_notify.insert(*subscription_id);
+                    } else {
+                        return Err(Error::MissingEventSubscription(*subscription_id));
+                    }
+                }
+            }
+
+            // Take note if a reconfiguration (new epoch) has occurred
+            if event_key == on_chain_config::new_epoch_event_key() {
+                reconfig_event_found = true;
+            }
+        }
+
+        // Notify event subscribers of the new events
+        for event_subscription_id in event_subscription_ids_to_notify {
+            if let Entry::Occupied(mut event_subscription) = self
+                .subscription_id_to_event_subscription
+                .entry(event_subscription_id)
+            {
+                event_subscription
+                    .get_mut()
+                    .notify_subscriber_of_events(version)?;
+            } else {
+                return Err(Error::MissingEventSubscription(event_subscription_id));
+            }
+        }
+
+        Ok(reconfig_event_found)
+    }
+
+    /// Reads the values of all on-chain configs at the specified version.
+    fn read_on_chain_configs(&self, version: Version) -> Result<OnChainConfigPayload, Error> {
+        // Get the access paths for all the configs in the registry
+        let config_access_paths = ON_CHAIN_CONFIG_REGISTRY
+            .iter()
+            .map(|config_id| config_id.access_path())
+            .collect();
+
+        // Read the config values
+        let on_chain_configs = self
+            .storage
+            .read()
+            .reader
+            .deref()
+            .batch_fetch_resources_by_version(config_access_paths, version)
+            .map_err(|error| {
+                Error::UnexpectedErrorEncountered(format!(
+                    "Failed to batch fetch resources at version: {:?}. Error: {:?}",
+                    version, error
+                ))
+            })?;
+
+        // Fetch the account state blob
+        let (account_state_blob, _) = self
+            .storage
+            .read()
+            .reader
+            .get_account_state_with_proof_by_version(config_address(), version)
+            .map_err(|error| {
+                Error::UnexpectedErrorEncountered(format!(
+                    "Failed to fetch account state with proof {:?}",
+                    error
+                ))
+            })?;
+        let account_state_blob = account_state_blob.ok_or_else(|| {
+            Error::UnexpectedErrorEncountered("Missing account state blob!".into())
+        })?;
+
+        // Fetch the new epoch from storage
+        let epoch = AccountState::try_from(&account_state_blob)
```

Comment:
nit: you could avoid looking up the account_state + epoch. iirc the reconfig event `NewEpochEvent` contains the epoch. could just pull that out before this call